File: ruby/sum-of-multiples/sum_of_multiples.rb
class SumOfMultiples

  def initialize(*factors)
    @factors = factors.empty? ? DEFAULT_FACTORS : factors
  end

  def to(limit)
    self.class.to(limit, @factors)
  end

  def self.to(limit, factors = DEFAULT_FACTORS)
    (1...limit).select { |number|
      factors.any? { |factor| number % factor == 0 }
    }.inject(0, &:+) # need 0 to satisfy "to(1)" case
  end

  private

  DEFAULT_FACTORS = [3, 5]

end

File: elixir/sum-of-multiples/sum_of_multiples.exs
defmodule SumOfMultiples do
  @doc """
  Adds up all numbers from 1 to a given end number that are multiples of the factors provided.
  """
  @spec to(non_neg_integer, [non_neg_integer]) :: non_neg_integer
  def to(limit, factors) do
    1..(limit-1) |> Enum.filter(&is_multiple(&1, factors)) |> Enum.sum
  end

  defp is_multiple(num, factors) do
    Enum.any?(factors, &(rem(num, &1) == 0))
  end

end
